### ISSUE TITLE
add webRTC filter via XSound to localAudioTrack, does not persist to remoteAudioTrack

### ISSUE THREAD

**Issue by cindyloo** [twilio/twilio-video.js#1443]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHxANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

here is my initial connection to Twilio (in useTwilio.ts):

```
        const stream = await (navigator as any).mediaDevices.getUserMedia({
          audio: true,
          video: true,
        });
        console.log("starting warp in useTwilio");
       
        audioMaskStart(stream, true); // XSound gets the AudioContext and connects some filters
      
        const lt = stream.getTracks(); // the output I hear confirms that is is warping my voice

        const room = await Video.connect(twilioToken, {
          name: roomName,
          tracks: lt,
          audio: true,
          video: true,
        });

        setRoom(room);
```

and here is the code that listens to participants joining
```
const [videoTracks, setVideoTracks] = useState<Video.VideoTrack[]>([]);
  const [audioTracks, setAudioTracks] = useState<Video.AudioTrack[]>([]);


useEffect(() => {
    console.log("audioTracks num are " + audioTracks.length);

    const audioTrack = audioTracks[0]; // note, only expects one
    //const currentRef = audioRef?.current;

if (audioTrack &&
        audioTrack.toString().indexOf("RemoteAudioTrack") > -1
      ) {
        const at = audioTrack as RemoteAudioTrack;

    const stream = new MediaStream([at.mediaStreamTrack]); // does it have to be cloned?

        const audioElement2 = document.createElement("audio");

        audioMaskStart(stream, true); // pass to XSound, which just connects filters to the stream
        audioElement2.srcObject = stream; // is this necessary if we attach? or does the attach not work in this case with updated tracks?
        document.getElementById("audio_div").appendChild(audioElement2);

        at.attach(audioElement2); // this removes and adds the this.mediaStreamTrack - how do we know this is the track as above?


},[audioTracks]; // updated when the participants join, tracks are then added and listened for publish event
 ```

**Expected behavior:**

any local stream filtering/modulating or behavior should be reflected in the track that is published to other subscribers (eg, participants in the room). A pitch-shifted voice should be pitch-shifted locally and remotely

**Actual behavior:**

the local audio track is modified and I can hear my shifted voice. It is also recorded into the Twilio audio track - although I will add that it sounds like it has recorded my voice AND my pitch-shifted voice.
However, this is not what any joining participant hears - they only hear my natural voice

my main question is, am I doing this right? is there a certain way I should be accessing the remoteTrack, or attaching it to an audio element (eg, I notice that in `attach`, all mediaStreamTracks are removed and `this.mediaStreamTrack` is added)? or is it more in the setup of the local track initially and how it is persisted through Twilio? or does it just not work for now? thanks for any help/pointers!! Could it be related to #1143 in that I am somehow not getting access to the right mediaStreamTrack?

logs:
```
starting warp in useTwilio
instrument.js:109 Starting audio masking
instrument.js:109 audioinput: Default - Internal Microphone (Built-in) id = default
instrument.js:109 audioinput: Internal Microphone (Built-in) id = 7c60247e55f8ea2bc812a3deb13df41b296a5423a4fb6fb7fb866a18f9a8b0ac
instrument.js:109 videoinput: FaceTime HD Camera id = 0ab71f83d1939671564afb013c9c38ba4b3f9eaed7d9f35916a505310f02c163
instrument.js:109 audiooutput: Default - Internal Speakers (Built-in) id = default
instrument.js:109 audiooutput: Internal Speakers (Built-in) id = f2a48f8d417dd0ee9bde7f608c8c0b9935559535945a4db248ab59977b739665
instrument.js:109 starting up with filters
instrument.js:109 {"mastervolume":1,"stereo":{"state":false,"time":0},"compressor":{"state":true,"threshold":-24,"knee":30,"ratio":12,"attack":0.003000000026077032,"release":0.25},"distortion":{"state":false,"curve":"clean","amount":0.5,"samples":4096,"drive":1,"color":350,"tone":350},"wah":{"state":false,"auto":false,"cutoff":350,"depth":0,"rate":0,"resonance":1},"pitchshifter":{"state":true,"pitch":0.85},"equalizer":{"state":true,"bass":-14,"middle":7,"treble":-20,"presence":-6},"filter":{"state":true,"type":"lowpass","frequency":350,"Q":1,"gain":0,"range":0.1,"attack":0.01,"decay":0.3,"sustain":1,"release":1},"autopanner":{"state":false,"depth":0,"rate":0},"tremolo":{"state":false,"depth":0,"rate":0,"wave":"sine"},"ringmodulator":{"state":false,"depth":1,"rate":0},"phaser":{"state":false,"stage":12,"frequency":350,"resonance":1,"depth":0,"rate":0,"mix":0,"feedback":0},"flanger":{"state":false,"time":0,"depth":0,"rate":0,"mix":0,"tone":350,"feedback":0},"chorus":{"state":false,"time":0,"depth":0,"rate":0,"mix":0,"tone":350,"feedback":0},"delay":{"state":false,"time":0,"dry":1,"wet":0,"tone":350,"feedback":0},"reverb":{"state":false,"dry":1,"wet":0,"tone":350},"panner":{"state":false,"positions":{"x":0,"y":0,"z":0},"orientations":{"x":1,"y":0,"z":0},"refDistance":1,"maxDistance":10000,"rolloffFactor":1,"coneInnerAngle":360,"coneOuterAngle":360,"coneOuterGain":0,"panningModel":"HRTF","distanceModel":"inverse"},"listener":{"state":true,"positions":{"x":0,"y":0,"z":0},"forwards":{"x":0,"y":0,"z":-1},"ups":{"x":0,"y":1,"z":0}},"envelopegenerator":{"attack":0.01,"decay":0.3,"sustain":0.5,"release":1},"stream":{"output":true,"noisegate":{"level":0},"noisesuppressor":{"threshold":0}}}
  audioTracks num are 1
instrument.js:109 audioTrack toString is [LocalAudioTrack #1: 777def26-23db-4b39-bd3f-014665681143]
instrument.js:109 audioTrack label is Default - Internal Microphone (Built-in)
instrument.js:109 audioTrack kind is audio



audioTracks num are 1
instrument.js:109 audioTrack toString is [RemoteAudioTrack #5: MTc00de26a208ef76ba3891bcf02868038]
instrument.js:109 audioTrack label is fbfa9339-5d51-433c-8618-5995646b2ff7
instrument.js:109 audioTrack kind is audio
instrument.js:109 Starting audio masking

instrument.js:109 starting up with filters
instrument.js:109 {"mastervolume":1,"stereo":{"state":false,"time":0},"compressor":{"state":true,"threshold":-24,"knee":30,"ratio":12,"attack":0.003000000026077032,"release":0.25},"distortion":{"state":false,"curve":"clean","amount":0.5,"samples":4096,"drive":1,"color":350,"tone":350},"wah":{"state":false,"auto":false,"cutoff":350,"depth":0,"rate":0,"resonance":1},"pitchshifter":{"state":true,"pitch":0.85},"equalizer":{"state":true,"bass":-14,"middle":7,"treble":-14,"presence":-6},"filter":{"state":true,"type":"lowpass","frequency":202.34375,"Q":1,"gain":0,"range":0.1,"attack":0.01,"decay":0.3,"sustain":1,"release":1},"autopanner":{"state":false,"depth":0,"rate":0},"tremolo":{"state":false,"depth":0,"rate":0,"wave":"sine"},"ringmodulator":{"state":false,"depth":1,"rate":0},"phaser":{"state":false,"stage":12,"frequency":350,"resonance":1,"depth":0,"rate":0,"mix":0,"feedback":0},"flanger":{"state":false,"time":0,"depth":0,"rate":0,"mix":0,"tone":350,"feedback":0},"chorus":{"state":false,"time":0,"depth":0,"rate":0,"mix":0,"tone":350,"feedback":0},"delay":{"state":false,"time":0,"dry":1,"wet":0,"tone":350,"feedback":0},"reverb":{"state":false,"dry":1,"wet":0,"tone":350},"panner":{"state":false,"positions":{"x":0,"y":0,"z":0},"orientations":{"x":1,"y":0,"z":0},"refDistance":1,"maxDistance":10000,"rolloffFactor":1,"coneInnerAngle":360,"coneOuterAngle":360,"coneOuterGain":0,"panningModel":"HRTF","distanceModel":"inverse"},"listener":{"state":true,"positions":{"x":0,"y":0,"z":0},"forwards":{"x":0,"y":0,"z":-1},"ups":{"x":0,"y":1,"z":0}},"envelopegenerator":{"attack":0.01,"decay":0.3,"sustain":0.5,"release":1},"stream":{"output":true,"noisegate":{"level":0},"noisesuppressor":{"threshold":0}}}

```
**Software versions:**

 - [ x] Browser(s): Chrome 88.0.43
 - [x ] Operating System: Mac Os 11.2.2
 - [ x] twilio-video.js: 2.7.2
 - [x ] Third-party libraries (e.g., Angular, React, etc.):  React 16.13.1

**Comment by charliesantos** [comment]
Hi @cindyloo ,

Thanks for reaching out! Let me gather more information and get back to you on this one. Meanwhile, have you tried other browsers? Is this happening only on Chrome?

Thank you,
Charlie

**Comment by cindyloo** [comment]
thanks @charliesantos! I did just try on Safari Version 14.0.3 and the remote track is also missing the voice filter

**Comment by cindyloo** [comment]
hey there @charliesantos - so I am trying to understand twilio-video and looking at possible places where the problem could be.

I turned on the debugging when connecting to Twilio. Is there anything and particular I could inspect? I see there is are codecs determined, planB vs unified... Would a good test be to try to stop or disable the localAudioTrack and restart it in case the signaling isn't updating?

let me know, I'm happy to get my hands dirty, just need some guidance :)

**Comment by spalmertwilio** [comment]
@charliesantos this is connected to support request 5965919 and VIDEO-4719

**Comment by charliesantos** [comment]
Thanks @spalmertwilio .

@cindyloo thanks for looking more into this! You can access the local audio track that is being published to the room. Once you are able to access it, can you please verify whether you can hear the expected output or not? That'll give us an idea where to look first - local or remote audio track.

```js
...// Your other code here

// Connect to the room
const room = await Video.connect(twilioToken, {...});

// Access local audio track.
// Please try to listen to this track and verify whether you can hear the expected output or not.
const audioTrack = Array.from(room.localParticipant.audioTracks.values())[0].track.mediaStreamTrack;
```

**Comment by cindyloo** [comment]
hi @charliesantos 
so, I have clarified some issues:
1) I had multiple audio elements and not all were muted. This made it very difficult to debug what was happening
2) once I reduced the number of audio elements to *just* the remote audio tracks, I hit on a couple of issues
a) I made two elements - one for the source remote track, and one for the destination track
b) I successfully could warp the destination track if I made it this way:
as I found (which surprisingly no one pointed me to) on[ stackoverflow from 2018](https://stackoverflow.com/questions/52512823/is-it-possible-to-do-voice-pitch-shifting-in-twilio-group-video)

```
const audioOutput = audioContext.createMediaStreamDestination();
localAudioTracksFromDest = new Video.LocalAudioTrack(
            audioOutput.stream.getAudioTracks()[0]
          );

const audioElementDest = localAudioTracksFromDest?.attach();
 document
          .getElementById(`audio_div${uniqueName}`)
          ?.appendChild(audioElementDest)
          .setAttribute("id", `audioDest${localOrRemote}${uniqueName}`);

```
it did not work for me this way (as suggested by Daniel in your customer support)
```
const stream = new MediaStream([track.mediaStreamTrack.clone()])
 const node = audioContext.createMediaStreamSource(stream);
const audioOutput = audioContext.destination;
node.connect(<a filter>)
<a filter>.connect(audioOutput);
 audioElement.srcObject = destination.stream;

```

**Comment by charliesantos** [comment]
Hi @cindyloo , thank you for the additional information. I will be looking more into this within the next 2 weeks and provide you a more detailed answer. I'll let you know if I need more information from you.

Just so I understand your use case better, you are adding some audio filters to the local audio track and wants to publish it. Then you expect all participants in the room to hear the audio with the applied filters correct?

**Comment by charliesantos** [comment]
Hi @cindyloo , just following up on my question above. Depending on your use case, we might have a different suggestion.
Also, audio processing is not currently supported by the SDK but it's something we are considering.

**Comment by cindyloo** [comment]
hi @charliesantos sorry for the delay, I was away
I was able to implement the voice pitchshift with the code above. I targeted the output stream of the remote audio of the participants. If a participant wants to mask his or her voice, I mark it and add the filter to the incoming remote stream.
I am also able to apply that pitchshift to the local voice, but the echo occurs unless they have headphones. I lowered the 'wet' setting to reduce the echo.

It would be great if adding additional filters was easier and supported.
